### PR TITLE
Automatically treat patients

### LINKED ISSUE #598: Personnel is automatically assigned to patients and treats them
When personnel and patients are in the same autonomous region, the personnel do triage and they patients and personnel are automatically matched such that the patients are being treated.

For this, we extend the existing tick action,

Requirement: M3b

### PR DESCRIPTION
**Pull request by lukasrad02**
Closes #598

### CODE REVIEW

File: shared/src/simulation/behaviors/treat-patients.ts [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+                                draftState.currentTime + behaviorState.interval
+                            )
+                        );
+                    }
+                    break;
+                case 'materialAvailableEvent':
+                case 'newPatientEvent':
+                case 'personnelAvailableEvent':
+                    Object.values(simulatedRegion.activities)
+                        .filter(
```
Comment (Nils1729) [inline shared/src/simulation/behaviors/treat-patients.ts:56]: I would rather the behavior saved the id of its activity in cases where they are strongly coupled like this to reduce surprising side effects.

File: shared/src/store/action-reducers/simulated-region.ts [hunk 2]
```diff
@@ -184,6 +205,47 @@ export namespace SimulatedRegionActionReducers {
...
+                if (element.type === 'vehicle') {
+                    sendSimulationEvent(
+                        simulatedRegion,
+                        VehicleArrivedEvent.create(
+                            element.id,
+                            simulatedRegion.id,
+                            draftState.currentTime
+                        )
+                    );
+                } else if (element.type === 'patient') {
```
Comment (Nils1729) [inline shared/src/store/action-reducers/simulated-region.ts:217]: A `switch`  statement might be better suited.

File: shared/src/simulation/activities/reassign-treatments.ts [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+    implements SimulationActivityState
+{
+    @IsValue('reassignTreatmentsActivity' as const)
+    public readonly type = 'reassignTreatmentsActivity';
+
+    @IsUUID(4, uuidValidationOptions)
+    public readonly id!: UUID;
+
+    @IsUUID(4, uuidValidationOptions)
+    public readonly behaviorId!: UUID;
```
Comment (Dassderdie) [inline shared/src/simulation/activities/reassign-treatments.ts:33]:
```suggestion
    @IsUUID(4, uuidValidationOptions)
    public readonly id: UUID;

    @IsUUID(4, uuidValidationOptions)
    public readonly behaviorId: UUID;
```

I believe you don't need these, as they are assigned in the constructor.

If you use the Non-null assertion operator here, you don't get a type error if you forget to assign it in the constructor.

Similar for all the other such classes.

File: shared/src/simulation/activities/reassign-treatments.ts [hunk 4]
```diff
@@ -0,0 +1,632 @@
...
+
+    const groupedPersonnel = groupBy(
+        cateringPersonnel,
+        (personnel) => personnel.personnel.personnelType
+    );
+
+    let securedPatients = 0;
+
+    groupedPatients.red?.forEach((patient) => {
+        const notSanResult = findAssignablePersonnel(
```
Comment (benn02) [inline shared/src/simulation/activities/reassign-treatments.ts:486]: I dont like the Name... maybe name it notSanFound or assignableNotSan something Same for the variables below

File: shared/src/simulation/activities/reassign-treatments.ts [hunk 5]
```diff
@@ -0,0 +1,632 @@
...
+    catersFor: CatersFor;
+}
+
+interface CateringPersonnel {
+    personnel: Mutable<Personnel>;
+    priority: number;
+    catersFor: CatersFor;
+}
+
+function createCateringMaterial(
```
Comment (Nils1729) [inline shared/src/simulation/activities/reassign-treatments.ts:179]:
```suggestion
function createCateringMaterials(
```

File: shared/src/simulation/activities/reassign-treatments.ts (hunk 4 above)
Comment (lukasrad02) [reply shared/src/simulation/activities/reassign-treatments.ts]: Did it in [0612587](https://github.com/hpi-sam/digital-fuesim-manv/pull/670/commits/06125872953c08fac27aedc0c38d5ec48832a110)

File: shared/src/simulation/utils/vehicle.ts [hunk 6]
```diff
@@ -26,13 +33,25 @@ export function unloadVehicle(
...
+                    sendSimulationEvent(
+                        simulatedRegion,
+                        MaterialAvailableEvent.create(element.id)
+                    );
+                } else if (element.type === 'personnel') {
+                    sendSimulationEvent(
+                        simulatedRegion,
+                        PersonnelAvailableEvent.create(element.id)
+                    );
+                }
```
[the remark this answers is not in the document]
Comment (lukasrad02) [reply shared/src/simulation/utils/vehicle.ts:64]: Fixed in https://github.com/hpi-sam/digital-fuesim-manv/pull/670/commits/b3e0009926d5542980f62aef5686c805b6a002ff

File: shared/src/simulation/activities/reassign-treatments.ts [hunk 7]
```diff
@@ -0,0 +1,632 @@
...
+            if (!leaderId) {
+                terminate();
+                return;
+            }
+
+            // The leader does not treat patients
+            personnel.splice(
+                personnel.findIndex((pers) => pers.id === leaderId),
+                1
+            );
```
[the remark this answers is not in the document]
Comment (lukasrad02) [reply shared/src/simulation/activities/reassign-treatments.ts]: Thank you for the hint! I've fixed it in https://github.com/hpi-sam/digital-fuesim-manv/pull/670/commits/3d513962f26562e94f6c2fc6a11dabb577a4bf98

File: shared/src/simulation/activities/reassign-treatments.ts [hunk 8]
```diff
@@ -0,0 +1,632 @@
...
+
+        if (!map[key]) {
+            map[key] = [];
+        }
+
+        map[key]!.push(element);
+    });
+
+    return map;
+}
```
[the remark this answers is not in the document]
Comment (lukasrad02) [reply shared/src/simulation/activities/reassign-treatments.ts]: Fixed in https://github.com/hpi-sam/digital-fuesim-manv/pull/670/commits/0d572015d795212e1dac431464cc1186f0203f24

File: shared/src/simulation/activities/reassign-treatments.ts (hunk 5 above)
Comment (lukasrad02) [reply shared/src/simulation/activities/reassign-treatments.ts]: Fixed in https://github.com/hpi-sam/digital-fuesim-manv/pull/670/commits/db87774da41bbe59bd45ef7f524865cf1a5619e4

File: shared/src/simulation/activities/reassign-treatments.ts [hunk 9]
```diff
@@ -388,7 +388,7 @@ function sumOfTreatments(cateringPersonnel: CateringPersonnel): number {
  */
 function findAssignablePersonnel(
     groupedPersonnel: {
-        [Key in PersonnelType]: CateringPersonnel[] | undefined;
+        [Key in PersonnelType]?: CateringPersonnel[] | undefined;
```
Comment (Nils1729) [inline shared/src/simulation/activities/reassign-treatments.ts:391]: Please remove `|undefined` if possible

**Review by benn02 — APPROVED** [review]
LGTM

**Pull request merged**